spark/src/main/java/org/apache/iceberg/spark/SparkDataFile.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.Row;
+
+public class SparkDataFile implements DataFile {
+
+  private final Type lowerBoundsType;
+  private final Type upperBoundsType;
+  private final Type keyMetadataType;
+  private final int fieldShift;
+  private final SparkStructLike wrappedPartition;
+  private Row wrapped;
+
+  public SparkDataFile(Types.StructType type) {
+    this.lowerBoundsType = type.fieldType("lower_bounds");
+    this.upperBoundsType = type.fieldType("upper_bounds");
+    this.keyMetadataType = type.fieldType("key_metadata");
+    this.wrappedPartition = new SparkStructLike(type.fieldType("partition").asStructType());
+    // the partition field is absent for unpartitioned tables
+    this.fieldShift = wrappedPartition.size() != 0 ? 1 : 0;
```

Comment:
As much as I don't like to use the index, referring by column name seems to incur an additional penalty as we will resolve the column name every time (based on what I see in Spark `Row`).

Reply:
If we assume that the type passed in here matches the Row, then we should be able to derive the ordinals from that type instead of hard-coding them. I think this would produce an int[] of expected offset to actual offset, like we use in the Avro generics. What do you think?

Reply:
Otherwise, I think it would probably be worth the performance penalty to ensure correctness.

Reply:
@rdblue, do you mean passing the Spark type alongside the Iceberg type and building a mapping in the constructor?

Reply:
Ideally, yes. But if that's not possible we know that the DataFile that will be passed in comes from a metadata table (because we control the implementations that use this) so it would be okay to assume that we know what the Spark schema is.

Reply:
Okay, I've updated this to handle arbitrary projections.